pkg/manifest/jsonmanifest/manifest.go

```diff
@@ -6,51 +6,94 @@ package jsonmanifest
 
 import (
 	"encoding/json"
+	"sync"
 
 	"github.com/ethersphere/bee/pkg/manifest"
 )
 
-// verify JSONManifest implements manifest.Interface.
-var _ manifest.Interface = (*JSONManifest)(nil)
+// verify jsonManifest implements manifest.Interface.
+var _ manifest.Interface = (*jsonManifest)(nil)
 
-// JSONManifest is a JSON representation of a manifest.
+// jsonManifest is a JSON representation of a manifest.
 // It stores manifest entries in a map based on string keys.
-type JSONManifest struct {
-	Entries map[string]*JSONEntry `json:"entries,omitempty"`
+type jsonManifest struct {
+	entriesMu sync.RWMutex // mutex for accessing the entries map
+	entries   map[string]*jsonEntry
 }
 
-// NewManifest creates a new JSONManifest struct and returns a pointer to it.
-func NewManifest() *JSONManifest {
-	return &JSONManifest{
-		Entries: make(map[string]*JSONEntry),
+// NewManifest creates a new jsonManifest struct and returns a pointer to it.
+func NewManifest() manifest.Interface {
+	return &jsonManifest{
+		entries: make(map[string]*jsonEntry),
 	}
 }
 
 // Add adds a manifest entry to the specified path.
-func (m *JSONManifest) Add(path string, entry manifest.Entry) {
-	m.Entries[path] = NewEntry(entry.Reference(), entry.Name(), entry.Headers())
+func (m *jsonManifest) Add(path string, entry manifest.Entry) {
+	m.entriesMu.Lock()
+	defer m.entriesMu.Unlock()
+
+	m.entries[path] = &jsonEntry{
+		reference: entry.Reference(),
+		name:      entry.Name(),
+		headers:   entry.Headers(),
+	}
 }
 
 // Remove removes a manifest entry on the specified path.
-func (m *JSONManifest) Remove(path string) {
-	delete(m.Entries, path)
+func (m *jsonManifest) Remove(path string) {
+	m.entriesMu.Lock()
+	defer m.entriesMu.Unlock()
+
+	delete(m.entries, path)
 }
 
 // Entry returns a manifest entry if one is found in the specified path.
-func (m *JSONManifest) Entry(path string) (manifest.Entry, error) {
-	if entry, ok := m.Entries[path]; ok {
-		return entry, nil
+func (m *jsonManifest) Entry(path string) (manifest.Entry, error) {
+	m.entriesMu.RLock()
+	defer m.entriesMu.RUnlock()
+
+	entry, ok := m.entries[path]
+	if !ok {
+		return nil, manifest.ErrNotFound
 	}
 
-	return nil, manifest.ErrNotFound
+	// return a copy to prevent external modification
+	return NewEntry(entry.Reference(), entry.Name(), entry.Headers()), nil
```

Comment:
not 100% on this, but since we are using pointer semantics, we should be careful not to have the manifest entries modified from outside, no?

Reply:
you can easily check if this is necessary or not with a unit test (create an entry, get it, change it, and get it again from the source to see if the instance was changed). this test should also probably be shipped with the code. I am not entirely sure though, we actually _want_ to allow editing of entries in this manner, but I would assume no for now.

Reply:
thanks, i added the `TestEntryModification` unit test. turns out we did need the copy, after all.

afaik, manifest entries should only be modifiable (as part of the manifest) through `manifest.Interface` or `manifest.Entry` functions, whether it is `Add` or some other eventual method.